File: Client/wwwroot/js/WebRTCClientTaker.ts
namespace SmartProctor {
    class ProctorConnection {
        public desktopConnection: RTCPeerConnection;
        public cameraConnection: RTCPeerConnection;
    }

    /**
     * Typescript implementation of the WebRTC related functions in the test-taker side
     */
    export class WebRTCClientTaker {
        private proctorConnections: { [userName: string]: ProctorConnection } = {};
        private desktopStream: MediaStream;
        private cameraStream: MediaStream;
        private desktopVideoElem: Element;
        private cameraVideoElem: Element;
        private rtcConfig: RTCConfiguration;
        private cameraCanvas: HTMLCanvasElement;
        private cameraImage;
        public helper;

        public async init(helper, iceServers: string[], proctors: string[]) {
            this.helper = helper;
            if (iceServers != null) {
                this.rtcConfig = {iceServers: [{urls: iceServers}]};
            } else {
                this.rtcConfig = null;
            }

            proctors.forEach((proctor) => {
                let desktopConn = new RTCPeerConnection(this.rtcConfig);
                let cameraConn = new RTCPeerConnection(this.rtcConfig); 

                desktopConn.onicecandidate = async (e) => {
                    await helper.invokeMethodAsync("_onDesktopIceCandidate", proctor, e.candidate);
                };

                desktopConn.onconnectionstatechange = async (e) => {
                    await helper.invokeMethodAsync("_onDesktopConnectionStateChange", proctor, desktopConn.connectionState);
                };
                
                cameraConn.onicecandidate = async (e) => {
                    await helper.invokeMethodAsync("_onCameraIceCandidate", proctor, e.candidate);
                };
                
                cameraConn.onconnectionstatechange= async (e) => {
                    await helper.invokeMethodAsync("_onCameraConnectionStateChange", proctor, cameraConn.connectionState);
                };
                
                this.proctorConnections[proctor] = {
                    desktopConnection: desktopConn,
                    cameraConnection: cameraConn
                };
            });
        }

        public async obtainDesktopStream(): Promise<string> {
            // @ts-ignore
            this.desktopStream = await navigator.mediaDevices.getDisplayMedia();
            // @ts-ignore
            this.desktopStream.oninactive = async (_) => {
                await this.helper.invokeMethodAsync("_onDesktopInactivated");
            };
            return this.desktopStream.getTracks()[0].label;
        }

        public async obtainCameraStream(mjpegUrl: string): Promise<void> {
            if (this.cameraCanvas == null) {
                // @ts-ignore
                this.cameraCanvas = document.createElement<HTMLCanvasElement>("canvas");
                this.cameraCanvas.width = 858;
                this.cameraCanvas.height = 480;
            }
            
            if (this.cameraImage == null) {
                this.cameraImage = new Image();
                this.cameraImage.crossOrigin = "anonymous";
            }
            
            this.cameraImage.src = mjpegUrl;
            // @ts-ignore
            this.cameraStream = this.cameraCanvas.captureStream();
            // @ts-ignore
            this.cameraStream.oninactive = async (_) => {
                await this.helper.invokeMethodAsync("_onCameraInactivated");
            };
            window.setInterval(() => {
                this.cameraCanvas.getContext("2d").drawImage(this.cameraImage, 0, 0);
            }, 15);
        }
        

        public async startStreaming() {
            for (let proctor in this.proctorConnections) {
                let conn = this.proctorConnections[proctor];
                this.desktopStream.getTracks().forEach((track) => {
                    conn.desktopConnection.addTrack(track, this.desktopStream);
                });

                this.cameraStream.getTracks().forEach((track) => {
                    conn.cameraConnection.addTrack(track, this.cameraStream);
                });

                let desktopOffer = await conn.desktopConnection.createOffer();
                await conn.desktopConnection.setLocalDescription(desktopOffer);
                
                let cameraOffer = await conn.cameraConnection.createOffer();
                await conn.cameraConnection.setLocalDescription(cameraOffer)

                // Send the local SDP through SignalR in .NET
                await this.helper.invokeMethodAsync("_onCameraSdp", proctor, cameraOffer);
                await this.helper.invokeMethodAsync("_onDesktopSdp", proctor, desktopOffer);
            }
        }

        public async reconnectToProctor(proctor: string) {
            let conn = this.proctorConnections[proctor];

            let desktopOffer = await conn.desktopConnection.createOffer();
            await conn.desktopConnection.setLocalDescription(desktopOffer);
            
            let cameraOffer = await conn.cameraConnection.createOffer();
            await conn.cameraConnection.setLocalDescription(cameraOffer);

            // Send the local SDP through SignalR in .NET
            await this.helper.invokeMethodAsync("_onDesktopSdp", proctor, desktopOffer);
            await this.helper.invokeMethodAsync("_onCameraSdp", proctor, cameraOffer);
        }

        public setDesktopVideoElement(elementId: string) {
            if (this.desktopVideoElem != null)
                // @ts-ignore
                this.desktopVideoElem.srcObject = null;
            this.desktopVideoElem = document.getElementById(elementId);
            // @ts-ignore
            this.desktopVideoElem.srcObject = this.desktopStream;
        }

        public setCameraVideoElement(elementId: string) {
            if (this.cameraVideoElem != null)
                // @ts-ignore
                this.cameraVideoElem.srcObject = null;
            this.cameraVideoElem = document.getElementById(elementId);
            // @ts-ignore
            this.cameraVideoElem.srcObject = this.cameraStream;
        }

        public async receivedDesktopAnswerSDP(proctor: string, sdp: RTCSessionDescriptionInit) {
            await this.proctorConnections[proctor].desktopConnection.setRemoteDescription(sdp);
        }

        public async receivedDesktopIceCandidate(proctor: string, candidate: RTCIceCandidate) {
            await this.proctorConnections[proctor].desktopConnection.addIceCandidate(candidate);
        }

        public async receivedCameraAnswerSDP(proctor: string, sdp: RTCSessionDescriptionInit) {
            await this.proctorConnections[proctor].cameraConnection.setRemoteDescription(sdp);
        }

        public async receivedCameraIceCandidate(proctor: string, candidate: RTCIceCandidate) {
            await this.proctorConnections[proctor].cameraConnection.addIceCandidate(candidate);
        }

        public async onProctorReconnected(proctor: string) {
            let conn = this.proctorConnections[proctor];

            let desktopOffer = await conn.desktopConnection.createOffer();
            await conn.desktopConnection.setLocalDescription(desktopOffer);

            let cameraOffer = await conn.cameraConnection.createOffer();
            await conn.cameraConnection.setLocalDescription(cameraOffer);

            // Send the local SDP through SignalR in .NET
            await this.helper.invokeMethodAsync("_onDesktopSdp", proctor, desktopOffer);
            await this.helper.invokeMethodAsync("_onCameraSdp", proctor, cameraOffer);
        }
    }

}

let webRTCClientTaker: SmartProctor.WebRTCClientTaker;

export function create(helper, iceServers: string[], proctors: string[]) {
    if (webRTCClientTaker == null) {
        webRTCClientTaker = new SmartProctor.WebRTCClientTaker();
        webRTCClientTaker.init(helper, iceServers, proctors);
    }

    return webRTCClientTaker;
}

File: Client/Interops/WebRTCClientTaker.cs
using System;
using System.Threading.Tasks;
using BrowserInterop.Extensions;
using Microsoft.JSInterop;
using SmartProctor.Shared.WebRTC;

namespace SmartProctor.Client.Interops
{
    public class WebRTCClientTaker
    {
        private IJSRuntime _jsRuntime;
        private IJSObjectReference _jsObj;

        private DotNetObjectReference<WebRTCClientTaker> _dotRef;
        private string[] _proctors;
        
        public event EventHandler<(string, RTCIceCandidate)> OnCameraIceCandidate;
        public event EventHandler<(string, RTCSessionDescriptionInit)> OnCameraSdp;
        public event EventHandler<(string, RTCIceCandidate)> OnDesktopIceCandidate;
        public event EventHandler<(string, RTCSessionDescriptionInit)> OnDesktopSdp;
        public event EventHandler<(string, string)> OnCameraConnectionStateChange;
        public event EventHandler<(string, string)> OnDesktopConnectionStateChange;
        public event EventHandler OnDesktopInactivated;
        public event EventHandler OnCameraInactivated;

        public WebRTCClientTaker(IJSRuntime jsRuntime, string[] proctors)
        {
            _jsRuntime = jsRuntime;
            _proctors = proctors;
        }

        private async ValueTask Init()
        {
            if (_jsObj == null)
            {
                _dotRef = DotNetObjectReference.Create<WebRTCClientTaker>(this);
                var module = await _jsRuntime.InvokeAsync<IJSObjectReference>(
                    "import", "./js/WebRTCClientTaker.js");
                _jsObj = await module.InvokeAsync<IJSObjectReference>("create", _dotRef, null, _proctors);
            }
        }

        public async ValueTask<string> ObtainDesktopStream()
        {
            await Init();
            return await _jsObj.InvokeAsync<string>("obtainDesktopStream");
        }

        public async ValueTask ObtainCameraStream(string mjpegUrl)
        {
            await Init();
            await _jsObj.InvokeVoidAsync("obtainCameraStream", mjpegUrl);
        }

        public async ValueTask StartStreaming()
        {
            await Init();
            await _jsObj.InvokeVoidAsync("startStreaming");
        }

        public async ValueTask ReconnectToProctor(string proctor)
        {
            await Init();
            await _jsObj.InvokeVoidAsync("reconnectToProctor");
        }

        public async ValueTask SetDesktopVideoElement(string elementId)
        {
            await Init();
            await _jsObj.InvokeVoidAsync("setDesktopVideoElement", elementId);
        }

        public async ValueTask SetCameraVideoElement(string elementId)
        {
            await Init();
            await _jsObj.InvokeVoidAsync("setCameraVideoElement", elementId);
        }

        public async ValueTask ReceivedDesktopAnswerSDP(string proctor, RTCSessionDescriptionInit sdp)
        {
            await Init();
            await _jsObj.InvokeVoidAsync("receivedDesktopAnswerSDP", proctor, sdp);
        }

        public async ValueTask ReceivedDesktopIceCandidate(string proctor, RTCIceCandidate candidate)
        {
            await Init();
            await _jsObj.InvokeVoidAsync("receivedDesktopIceCandidate", proctor, candidate);
        }
        
        public async ValueTask ReceivedCameraAnswerSDP(string proctor, RTCSessionDescriptionInit sdp)
        {
            await Init();
            await _jsObj.InvokeVoidAsync("receivedCameraAnswerSDP", proctor, sdp);
        }

        public async ValueTask ReceivedCameraIceCandidate(string proctor, RTCIceCandidate candidate)
        {
            await Init();
            await _jsObj.InvokeVoidAsync("receivedCameraIceCandidate", proctor, candidate);
        }
        
        [JSInvokable]
        public ValueTask _onCameraIceCandidate(string proctor, RTCIceCandidate candidate)
        {
            OnCameraIceCandidate?.Invoke(this, (proctor, candidate));
            return ValueTask.CompletedTask;
        }
        
        [JSInvokable]
        public ValueTask _onDesktopIceCandidate(string proctor, RTCIceCandidate candidate)
        {
            OnDesktopIceCandidate?.Invoke(this, (proctor, candidate));
            return ValueTask.CompletedTask;
        }

        [JSInvokable]
        public ValueTask _onCameraSdp(string proctor, RTCSessionDescriptionInit sdp)
        {
            OnCameraSdp?.Invoke(this, (proctor, sdp));
            return ValueTask.CompletedTask;
        }
        
        [JSInvokable]
        public ValueTask _onDesktopSdp(string proctor, RTCSessionDescriptionInit sdp)
        {
            OnDesktopSdp?.Invoke(this, (proctor, sdp));
            return ValueTask.CompletedTask;
        }

        [JSInvokable]
        public ValueTask _onCameraConnectionStateChange(string proctor, string connectionState)
        {
            OnCameraConnectionStateChange?.Invoke(this, (proctor, connectionState));
            return ValueTask.CompletedTask;
        }
        
        [JSInvokable]
        public ValueTask _onDesktopConnectionStateChange(string proctor, string connectionState)
        {
            OnDesktopConnectionStateChange?.Invoke(this, (proctor, connectionState));
            return ValueTask.CompletedTask;
        }

        [JSInvokable]
        public ValueTask _onDesktopInactivated()
        {
            OnDesktopInactivated?.Invoke(this, EventArgs.Empty);
            return ValueTask.CompletedTask;
        }

        [JSInvokable]
        public ValueTask _onCameraInactivated()
        {
            OnCameraInactivated?.Invoke(this, EventArgs.Empty);
            return ValueTask.CompletedTask;
        }
    }
}